### PR TITLE
solution first part of hw-internet-shop

### PR DESCRIPTION
**Pull request by RomanHreits**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -8,5 +8,68 @@
...
+                    <configuration>
+                        <source>${jdk.version}</source>
+                        <target>${jdk.version}</target>
+                        <encoding>${project.build.sourceEncoding}</encoding>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+    </build>
 </project>
```
Comment (aleksandr-hrankin) [inline pom.xml]: Лучше пустую строку добавить)

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        IntStream.range(0, Storage.products.size())
+                .filter(i -> Storage.products.get(i).getId().equals(product.getId()))
+                .forEach(i -> Storage.products.set(i, product));
+        return product;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        int size = Storage.products.size();
+        Storage.products.removeIf(n -> n.getId().equals(id));
```
Comment (aleksandr-hrankin) [inline src/main/java/com/internet/shop/dao/ProductDaoImpl.java]: Метод removeIfвозвращает boolean

File: src/main/java/com/internet/shop/db/Storage.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.internet.shop.model.Product;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Storage {
+    public static final List<Product> products = new ArrayList<>();
+    private static Long idGenerator = 0L;
+
+    public static Product addProduct(Product product) {
```
Comment (aleksandr-hrankin) [inline src/main/java/com/internet/shop/db/Storage.java:11]: Думаю можно сделать метод void как в примере)

Comment (RomanHreits) [reply src/main/java/com/internet/shop/db/Storage.java:11]: не знаю настільки правильно, але метод в дао add має вертати створений product із згенерованим id...тому тут ми його якраз і вертаємо

File: pom.xml [hunk 4]
```diff
@@ -8,5 +8,68 @@
...
 
-</project>
\ No newline at end of file
+    <dependencies>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (KseniiaMakarova) [inline pom.xml]: let's remove this dependency, we don't use it at the moment

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
+package com.internet.shop.dao;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/ProductDaoImpl.java]:
```suggestion
package com.internet.shop.dao.impl;
```

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    public Product update(Product product) {
+        IntStream.range(0, Storage.products.size())
+                .filter(i -> Storage.products.get(i).getId().equals(product.getId()))
+                .forEach(i -> Storage.products.set(i, product));
+        return product;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+       return Storage.products.removeIf(n -> n.getId().equals(id));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/ProductDaoImpl.java]: `n` is not informative

File: src/main/java/com/internet/shop/db/Storage.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package com.internet.shop.db;
+
+import com.internet.shop.model.Product;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Storage {
+    public static final List<Product> products = new ArrayList<>();
+    private static Long idGenerator = 0L;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/db/Storage.java]:
```suggestion
    private static Long productId = 0L;
```

File: src/main/java/com/internet/shop/db/Storage.java (hunk 3 above)
Comment (KseniiaMakarova) [reply src/main/java/com/internet/shop/db/Storage.java:11]: I think it is possible to return Product here

File: src/main/java/com/internet/shop/lib/Service.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package com.internet.shop.lib;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/lib/Service.java:8]: you also use `@Target` everywhere

File: src/main/java/com/internet/shop/model/Order.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
+package com.internet.shop.model;
+
+import java.util.List;
+import java.util.Objects;
+
+public class Order {
+    private Long id;
+    private List<Product> products;
+    private double price;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/Order.java]: take a look at the UML please

File: src/main/java/com/internet/shop/model/Order.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+    public Order(List<Product> products, double price) {
+        this.products = products;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/Order.java]: we will not get products from outside

File: src/main/java/com/internet/shop/model/Order.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        this.price = price;
+    }
+
+    public Long getId() {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/Order.java:17]: let's either delete lombok dependency from pom.xml or use it

File: src/main/java/com/internet/shop/model/ShoppingCart.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.List;
+import java.util.Objects;
+
+public class ShoppingCart {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
+
+    public ShoppingCart(List<Product> products, Long userId) {
+        this.products = products;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/ShoppingCart.java]: same - we will not receive products from outside

File: src/main/java/com/internet/shop/service/ProductServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
+package com.internet.shop.service;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/ProductServiceImpl.java]:
```suggestion
package com.internet.shop.service.impl;
```

File: src/main/java/com/internet/shop/service/ProductServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+import com.internet.shop.dao.ProductDao;
+import com.internet.shop.lib.Inject;
+import com.internet.shop.lib.Service;
+import com.internet.shop.model.Product;
+import java.util.List;
+
+@Service
+public class ProductServiceImpl implements ProductService {
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/ProductServiceImpl.java]:
```suggestion
```

File: src/main/java/com/internet/shop/service/ProductServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    private ProductDao productDao;
+
+    @Override
+    public Product create(Product product) {
+        return productDao.create(product);
+    }
+
+    @Override
+    public Product get(Long id) {
+        return productDao.get(id).orElseThrow();
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/ProductServiceImpl.java]: let's use `get()` for now

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 4 above)
Comment (RomanHreits) [reply pom.xml]: done

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java (hunk 5 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/dao/ProductDaoImpl.java]: done

File: src/main/java/com/internet/shop/dao/ProductDaoImpl.java (hunk 6 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/dao/ProductDaoImpl.java]: done

File: src/main/java/com/internet/shop/db/Storage.java (hunk 7 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/db/Storage.java]: done

File: src/main/java/com/internet/shop/lib/Service.java (hunk 8 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/lib/Service.java:8]: done

File: src/main/java/com/internet/shop/model/Order.java (hunk 9 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/model/Order.java]: done

File: src/main/java/com/internet/shop/model/Order.java (hunk 10 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/model/Order.java]: done

File: src/main/java/com/internet/shop/model/Order.java (hunk 11 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/model/Order.java:17]: deleted...done

File: src/main/java/com/internet/shop/model/ShoppingCart.java (hunk 12 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/model/ShoppingCart.java]: done

File: src/main/java/com/internet/shop/service/ProductServiceImpl.java (hunk 13 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/service/ProductServiceImpl.java]: done

File: src/main/java/com/internet/shop/service/ProductServiceImpl.java (hunk 14 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/service/ProductServiceImpl.java]: done

File: src/main/java/com/internet/shop/service/ProductServiceImpl.java (hunk 15 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/service/ProductServiceImpl.java]: done

File: src/main/java/com/internet/shop/dao/impl/ProductDao.java [hunk 16]
```diff
@@ -0,0 +1,17 @@
+package com.internet.shop.dao.impl;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDao.java]:
```suggestion
package com.internet.shop.dao;
```
interfaces should be in `dao` folder
their implementations - in `dao.impl` subfolder

File: src/main/java/com/internet/shop/service/impl/ProductService.java [hunk 17]
```diff
@@ -0,0 +1,16 @@
+package com.internet.shop.service.impl;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/impl/ProductService.java]:
```suggestion
package com.internet.shop.service;
```
same here

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/internet/shop/dao/impl/ProductDao.java (hunk 16 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/dao/impl/ProductDao.java]: done

File: src/main/java/com/internet/shop/service/impl/ProductService.java (hunk 17 above)
Comment (RomanHreits) [reply src/main/java/com/internet/shop/service/impl/ProductService.java]: done

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**